### PR TITLE
Address encoding cleanup

### PR DESCRIPTION
**Pull request by sipa**
This PR contains some of the changes left as TODO in #11167 (and built on top of that PR). They are not intended for backporting.

This removes the `CBase58`, `CBitcoinSecret`, `CBitcoinExtKey`, and `CBitcoinExtPubKey` classes, in favor of simple `Encode`/`Decode` functions. Furthermore, all Bitcoin-specific logic (addresses, WIF, BIP32) is moved to `key_io.{h,cpp}`, leaving `base58.{h,cpp}` as a pure utility that implements the base58 encoding/decoding logic.

### CODE REVIEW

**Comment by gmaxwell** [comment]
Should this perhaps be merged only after the rest of the backport intended segwit stuff is done?

**Comment by sipa** [comment]
Perhaps, yes. This is not urgent in any way.

File: src/base58.cpp [hunk 1]
```diff
@@ -313,39 +313,35 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
...
+    std::vector<unsigned char> data;
+    if (DecodeBase58Check(str, data)) {
+        const std::vector<unsigned char>& privkey_prefix = Params().Base58Prefix(CChainParams::SECRET_KEY);
+        if ((data.size() == 32 + privkey_prefix.size() || (data.size() == 33 + privkey_prefix.size() && data.back() == 1)) &&
+            std::equal(privkey_prefix.begin(), privkey_prefix.end(), data.begin())) {
+            bool compressed = data.size() == 33 + privkey_prefix.size();
+            key.Set(data.begin() + privkey_prefix.size(), data.begin() + privkey_prefix.size() + 32, compressed);
+        }
+    }
+    memory_cleanse(data.data(), data.size());
```
Comment (promag) [inline src/base58.cpp]: Important: IMO `DecodeBase58Check`should *cleanse* `data`before returning `false`.

Comment (promag) [reply src/base58.cpp]: Extra: In that case this function could be:
```cpp
    CKey key;
    std::vector<unsigned char> data;
    if (!DecodeBase58Check(str, data)) return key;

    const auto& prefix = Params().Base58Prefix(CChainParams::SECRET_KEY);
    int size = data.size() - prefix.size();
    if (size == 32 || (size == 33 && data.back() == 1)) {
        if (std::equal(prefix.begin(), prefix.end(), data.begin())) {
            ...
        }
    }

    memory_cleanse(data.data(), data.size());
    return key;
```

Comment (promag) [reply src/base58.cpp]: Side note: Another option is to create a PR to use custom allocator for all containers that are `memory_cleanse`.

**Review by promag — COMMENTED** [review]
See collapsed comment.

**Comment by promag** [comment]
In commit `Split key_io (address/key encodings) off from base58` move tests `{Decode,Encode}Destination` tests from base58_tests.cpp to `key_io_tests.cpp`?

**Comment by sipa** [comment]
@promag Added a commit that splits off the key_io tests from base58_tests.

**Comment by promag** [comment]
Needs rebase.

**Comment by sipa** [comment]
Rebased.

**Comment by laanwj** [comment]
Nice, I like how this gets rid of awkward conversion classes like CBitcoinSecret and replaces them with simple functions.
utACK.68f0c52

**Comment by sipa** [comment]
Rebased.

**Comment by sipa** [comment]
Rebased.

**Comment by sipa** [comment]
Rebased.

File: src/test/key_tests.cpp [hunk 2]
```diff
@@ -32,21 +32,16 @@ BOOST_FIXTURE_TEST_SUITE(key_tests, BasicTestingSetup)
...
-    CBitcoinSecret bsecret1, bsecret2, bsecret1C, bsecret2C, baddress1;
-    BOOST_CHECK( bsecret1.SetString (strSecret1));
-    BOOST_CHECK( bsecret2.SetString (strSecret2));
-    BOOST_CHECK( bsecret1C.SetString(strSecret1C));
-    BOOST_CHECK( bsecret2C.SetString(strSecret2C));
-    BOOST_CHECK(!baddress1.SetString(strAddressBad));
-
-    CKey key1  = bsecret1.GetKey();
+    CKey key1  = DecodeSecret(strSecret1);
     BOOST_CHECK(key1.IsCompressed() == false);
```
Comment (MarcoFalke) [inline src/test/key_tests.cpp]: This is initialized with false. I assume you'd also have to check for `key1.IsValid()`?

File: src/test/key_tests.cpp [hunk 3]
```diff
@@ -32,21 +32,16 @@ BOOST_FIXTURE_TEST_SUITE(key_tests, BasicTestingSetup)
...
-    CKey key2  = bsecret2.GetKey();
+    CKey key2  = DecodeSecret(strSecret2);
     BOOST_CHECK(key2.IsCompressed() == false);
-    CKey key1C = bsecret1C.GetKey();
+    CKey key1C = DecodeSecret(strSecret1C);
     BOOST_CHECK(key1C.IsCompressed() == true);
-    CKey key2C = bsecret2C.GetKey();
+    CKey key2C = DecodeSecret(strSecret2C);
     BOOST_CHECK(key2C.IsCompressed() == true);
+    CKey bad_address = DecodeSecret(strAddressBad);
```
Comment (MarcoFalke) [inline src/test/key_tests.cpp]: nit: `bad_key`

File: src/test/key_tests.cpp (hunk 2 above)
Comment (sipa) [reply src/test/key_tests.cpp]: Fixed.

File: src/test/key_tests.cpp (hunk 3 above)
Comment (sipa) [reply src/test/key_tests.cpp]: Fixed.

**Comment by sipa** [comment]
Rebased and addressed @MarcoFalke's comments.

**Review by maflcko — APPROVED** [review]
utACK 5f8d5f7d0e7c97e2f24aeb2fa0c8f60e26e29ae2. Just one naming-nit

**Comment by MarcoFalke** [comment]
re-utACK 2333475531 (was rebased with no other code changes)

**Comment by sipa** [comment]
Rebased.

**Comment by MarcoFalke** [comment]
re-utACK 92f1f8b

**Pull request merged**